src/backend_pgsql.c

```diff
@@ -260,8 +260,12 @@ backend_authenticate(const char *service, const char *user, const char *passwd,
 			rc = PAM_USER_UNKNOWN;
 		} else if (!PQgetisnull(res, 0, 0)) {
 			char *stored_pw = PQgetvalue(res, 0, 0);
-			if (!strcmp(stored_pw, (tmp = password_encrypt(options, user, passwd, stored_pw)))) rc = PAM_SUCCESS;
-			free (tmp);
+			if (options->pw_type == PW_FUNCTION) {
+				if (!strcmp(stored_pw, "t")) { rc = PAM_SUCCESS; }
+			} else {
```

Comment:
The point is to let the database handle password authentication (hashing) by using a function like the one below:
`CREATE FUNCTION auth(p_username IN TEXT,p_password IN TEXT) RETURNS BOOLEAN AS ...`
If the function returns TRUE the user is considered to be authenticated otherwise the pam module should deny permissions. In that case the auth query should look like:
`auth_query = SELECT auth(%u,%p)`